src/vdaf.rs

```diff
@@ -1,9 +1,229 @@
+    ) -> Result<Self::AggregateShare, VdafError>;
+}
+
+/// The Collector's role in the execution of a VDAF.
+pub trait Collector: Vdaf {
+    /// Combines aggregate shares into the aggregate result.
+    fn unshard<M: IntoIterator<Item = Self::AggregateShare>>(
+        &self,
+        agg_param: &Self::AggregationParam,
+        agg_shares: M,
+    ) -> Result<Self::AggregateResult, VdafError>;
+}
+
+/// Intermediate state of an Aggregator during the Prepare process.
+pub trait PrepareStep<O>: Sized {
+    /// The type of input messages.
+    type Input;
+
+    /// The type of output message.
+    type Output;
+
+    /// Compute the next state transition from the current state and the previous round of input
+    /// messages.
+    fn step<M: IntoIterator<Item = Self::Input>>(self, inputs: M) -> PrepareTransition<Self, O>;
+
+    /// Compute the Aggregator's first message.
+    fn start(self) -> Result<(Self, Self::Output), VdafError> {
+        match self.step(std::iter::empty()) {
+            PrepareTransition::Continue(new_state, output) => Ok((new_state, output)),
+            PrepareTransition::Fail(err) => Err(err),
+            PrepareTransition::Finish(_) => Err(VdafError::Uncategorized(
+                "start() resulted in early Finish transition".to_string(),
+            )),
+        }
+    }
+
+    /// Compute the Aggregator's next message from the previous round of messages.
+    fn next<M: IntoIterator<Item = Self::Input>>(
+        self,
+        inputs: M,
+    ) -> Result<(Self, Self::Output), VdafError> {
+        match self.step(inputs) {
+            PrepareTransition::Continue(new_state, output) => Ok((new_state, output)),
+            PrepareTransition::Fail(err) => Err(err),
+            PrepareTransition::Finish(_) => Err(VdafError::Uncategorized(
+                "next() resulted in early Finish transition".to_string(),
+            )),
+        }
+    }
+
+    /// Recover the Aggregator's output share.
+    fn finish<M: IntoIterator<Item = Self::Input>>(self, inputs: M) -> Result<O, VdafError> {
+        match self.step(inputs) {
+            PrepareTransition::Continue(_, _) => Err(VdafError::Uncategorized(
+                "finish() resulted in Continue transition".to_string(),
+            )),
+            PrepareTransition::Fail(err) => Err(err),
+            PrepareTransition::Finish(output_share) => Ok(output_share),
+        }
+    }
+}
+
+/// A state transition of an Aggregator during the Prepare process.
+pub enum PrepareTransition<S: PrepareStep<O>, O> {
+    /// Continue processing.
+    Continue(S, S::Output),
+
+    /// Finish processing and return the output share.
+    Finish(O),
+
+    /// Fail and return an error.
+    Fail(VdafError),
+}
+
+/// An aggregate share that can merged with aggregate shares of the same type.
+pub trait Aggregatable {
+    /// Update an aggregate share by merging it with another (`agg_share`).
+    fn merge(&mut self, agg_share: &Self) -> Result<(), VdafError>;
```

Comment:
I'd be cool with this. Just a small nit: I'd move the `lhs` into `merge` so that you can consume it and avoid a copy.

Reply:
Well, I might not want to move `lhs`. Anyway, this is premature because I can't show you the use case for either yet. Let's stick with `merge` for now, and I may make a PR on this when i adopt it.